### PR TITLE
feat: state: Fast migration for v15

### LINKED ISSUE #7870: v7 migration integration
integrate the actor v7 migration mechansim

### PR DESCRIPTION
**Pull request by arajasek**
## Related Issues

Fixes #7870 

## Checklist

Before you mark the PR ready for review, please make sure that:
- [x] All commits have a clear commit message.
- [x] The PR title is in the form of of `<PR type>: <#issue number> <area>: <change being made>`
    - example: ` fix: #1234 mempool: Introduce a cache for valid signatures`
    - `PR type`: _fix_, _feat_, _INTERFACE BREAKING CHANGE_, _CONSENSUS BREAKING_, _build_, _chore_, _ci_, _docs_, _misc_,_perf_, _refactor_, _revert_, _style_, _test_
    - `area`: _api_, _chain_, _state_, _vm_, _data transfer_, _market_, _mempool_, _message_, _block production_, _multisig_, _networking_, _paychan_, _proving_, _sealing_, _wallet_
- [ ] This PR has tests for new functionality or change in behaviour
- [ ] If new user-facing features are introduced, clear usage guidelines and / or documentation updates should be included in https://lotus.filecoin.io or [Discussion Tutorials.](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] CI is green

### CODE REVIEW

**Comment by arajasek** [comment]
Some more experimenting and tweaking is needed, but this gets the job done well. Will share latest numbers soon (roughly 35 mins for premigration, 3 mins for migration on my box).

File: chain/consensus/filcns/upgrades.go [hunk 1]
```diff
@@ -1245,8 +1247,15 @@ func PreUpgradeActorsV7(ctx context.Context, sm *stmgr.StateManager, cache stmgr
 		workerCount /= 2
 	}
 
-	config := nv15.Config{MaxWorkers: uint(workerCount)}
-	_, err := upgradeActorsV7Common(ctx, sm, cache, root, epoch, ts, config)
+	lbts, lbRoot, err := stmgr.GetLookbackTipSetForRound(ctx, sm, ts, epoch)
```
Comment (arajasek) [inline chain/consensus/filcns/upgrades.go:1250]: This migration requires us to be reorg-insensitive, so the premigration has to be based on finality.

File: chain/consensus/filcns/upgrades.go [hunk 2]
```diff
@@ -1255,7 +1264,12 @@ func upgradeActorsV7Common(
 	root cid.Cid, epoch abi.ChainEpoch, ts *types.TipSet,
 	config nv15.Config,
 ) (cid.Cid, error) {
-	buf := blockstore.NewTieredBstore(sm.ChainStore().StateBlockstore(), blockstore.NewMemorySync())
+	writeStore, err := autobatch.NewBlockstore(sm.ChainStore().StateBlockstore(), blockstore.NewMemorySync(), 100_000, 100, true)
```
Comment (arajasek) [inline chain/consensus/filcns/upgrades.go:1267]: 100_000 here wasn't chosen very scientifically...

**Comment by arajasek** [comment]
Update to the tagged actors RC before landing this.

File: chain/consensus/filcns/upgrades.go (hunk 2 above)
Comment (Stebalien) [reply chain/consensus/filcns/upgrades.go:1267]: This blockstore:

1. Doesn't guarantee a flush order.
2. Doesn't actually wait in `Flush` for the all data to be written.
3. Definitely wasn't written for this use-case.

3 isn't really an issue, but 1 and 2 are. Ideally the node won't be borked if you crash during a pre-migration.

**Review by Stebalien — CHANGES_REQUESTED** [review]

**Comment by Stebalien** [comment]
> 3 isn't really an issue, but 1 and 2 are. Ideally the node won't be borked if you crash during a pre-migration.

I'd just implement a much simpler version of this. Basically, you need something with 3 stages:

1. Buffered.
2. Flushing.
3. Backing.

Writes go into buffered. When it fills, `put` signals (uses a length-one channel) some background goroutine to flush. That background goroutine will move everything from buffered to flushing, and start writing it to the backing blockstore.

Gets check buffered, then flushing, then backing.

The background goroutine ensures that data is _always_ flushed in-order.

**Comment by arajasek** [comment]
Superseded by #7933

**Pull request closed without merging**